tests/Multipart/UploadStateTest.php

```diff
@@ -70,4 +70,60 @@ public function testSerializationWorks()
         $this->assertTrue($newState->isInitiated());
         $this->assertArrayHasKey('foo', $newState->getId());
     }
-}
+    /**
+     * @dataProvider getDisplayProgressCases
+     */
+    public function testDisplayProgressPrints(
+        $totalSize,
+        $totalUploaded,
+        $progressBar
+    ) {
+        $state = new UploadState([]);
+        $state->setProgressThresholds($totalSize);
+        $state->displayProgress($totalUploaded);
+
+        $this->expectOutputString($progressBar);
+    }
+
+    public function getDisplayProgressCases()
+    {
+        $progressBar = [
+            "Transfer initiated...\n|                    | 0.0%\n",
+            "|==                  | 12.5%\n",
+            "|=====               | 25.0%\n",
+            "|=======             | 37.5%\n",
+            "|==========          | 50.0%\n",
+            "|============        | 62.5%\n",
+            "|===============     | 75.0%\n",
+            "|=================   | 87.5%\n",
+            "|====================| 100.0%\nTransfer complete!\n"
+        ];
+        return [
+            [100000, 0, $progressBar[0]],
+            [100000, 12499, $progressBar[0]],
+            [100000, 12500, "$progressBar[0]$progressBar[1]"],
```

Comment:
could you add some more isolated/incremental cases to ensure we're hitting each threshold as expected?